### DISCUSSION TITLE
Is there a need for ADL serializer macros?

### DISCUSSION THREAD

**Discussion by SergejAnZu** [nlohmann/json#4008] [Q&A]
In our project we have quite a few non-default-constructible models, and writing a custom ADL serializer for each of them is quite annoying.

We came up with a macro that generates adl serializer sepcializations as stated in https://json.nlohmann.me/features/arbitrary_types/#how-can-i-use-get-for-non-default-constructiblenon-copyable-types

Compiler used: gcc version 10.2.0
C++ version: C++20

Before I invest time into setting up a proper contribution, I'd rather ask: 
- Is there a need for macro helpers supporting non-default-constructible types
- If not, what is the design decision behind that? 

Here's the snippet in question: 
```cpp
/**
 * @brief deserialize single value with appropriate type. Mandatory for JSON_DEFINE_TYPE_NON_DEFAULT
 *
 * @see NLOHMANN_JSON_FROM
 * @see JSON_DEFINE_TYPE_NON_DEFAULT
 */
#define JSON_FROM_DECLTYPE(v1) nlohmann_json_j.at(#v1).get<decltype(__type::v1)>(),

/**
 * @brief defines an ADL serializer for non-default-constructible types as defined in
 * https://json.nlohmann.me/features/arbitrary_types/#how-can-i-use-get-for-non-default-constructiblenon-copyable-types
 */
#define JSON_DEFINE_TYPE_NON_DEFAULT(Type, ...) \
  NLOHMANN_JSON_NAMESPACE_BEGIN \
  template <> \
  struct adl_serializer<Type> \
  { \
    using __type = Type; \
    static Type from_json(const json& nlohmann_json_j) \
    { \
      return {NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(JSON_FROM_DECLTYPE, __VA_ARGS__))}; \
    } \
    static void to_json(json& nlohmann_json_j, const Type& nlohmann_json_t) \
    { \
      NLOHMANN_JSON_EXPAND(NLOHMANN_JSON_PASTE(NLOHMANN_JSON_TO, __VA_ARGS__)) \
    } \
  }; \
  NLOHMANN_JSON_NAMESPACE_END
```

**Comment by nlohmann** [comment]
Looking into this.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
